Requirements.md

```diff
@@ -0,0 +1,101 @@
+# Requirements for GC
+
+This document identifies the key requirements for the managed memory proposal. Below we refer to the version of WASM with managed memory as being WASM-GC.
+
+## Goals
+
+1. Support for host-provided GC
+
+   Allow language implementers to use host-provided GC features as their primary means of supporting memory management. 
+   
+   On the whole, it would be preferable if language implementers could target WASM-GC as a natural target for their language. While it is difficult to be precise about this, an ideal scenario would be that language implementers could target WASM-GC as though it were just another back-end. In particular, targeting WASM-GC should try to avoid requiring changes to earlier stages of a compiler - such as the type checker or source-level transformations.
+   
+### Non-Goals
+1. Support for linear memory GC.
+
+   We do not attempt with this proposal to support alternate memory management strategies that rely on collecting linear memory.
+   
+## Criteria
+
+A criterion is intended to act as a metric of a 'soft value', such as 'desireability' or 'implementability'. Criteria are especially important when weighing the relative merits of features and in determining the importance that some scenario is satisfied.
+
+1. Prioriterize popular languages.
+   When deciding which features to provide it is important to place emphasis on features that are more likely to be used. We prioriterize support for the popular programming languages; while it is difficult to be precise about this, one measure is a public index such as the TIOBE index. The TIOBE ranking of the top 12 languages with managed memory (from the top 20 languages) is Java, Python, C#, Visual Basic, PHP, SQL, Swift, Go, Ruby, MATLAB, Perl, Scratch.
```

Comment:
TIOBE is probably pseudoscience, and in any case popularity is a very coarse metric since we don't know anything about the context for the popularity.  No doubt Java and C# and Ruby and PHP and Go are popular, but as server languages, sometimes on application servers where there may be no good options.  More relevant seems to be whether the language is a good fit for the client-side web (yes there are other domains, but it might be good to focus), whether there is existing code in the language that will be ported to the web (including libraries), and whether new code might be written in the language for the web.

Reply:
For better or worse, Perl, Swift and Matlab use RC. So do PHP and Python, at least primarily, although they have fallback GCs to deal with cycles. It's not clear if either of them would easily benefit from Wasm GC. Visual Basic is a sibling CLR language to C#. So the above list can probably be thinned to 3 relevant representatives: Java/JVM, C#/CLR, Go. That misses out on many other GC'ed languages. I'm surprised JS is not on the list, though.

Reply:
PHP has destructors on objects which must be called at the expected times, so would need to use explicit ref counting to preserve language semantics. Relying on Wasm GC to handle cycle collection wouldn't work either as it would not have a way to call destructors (no finalization support is planned yet).

JS seems mostly implementable on the various plans I've seen except for `WeakMap` which requires weak references that no proposal has included yet, and typed arrays which cannot perform fast multi-byte accesses using any proposal I've seen yet. (However implementing JS in a fast way requires run-time profiling, tiered JIT compilation, and a way to hack up some kind of on-stack replacement. My implementation sketch plans only for a baseline compiler with expected middling performance.)

Reply:
Erlang uses a mix of per Erlang process (green thread) heaps that are individually collected, but for large shared binaries (byte vectors), they are stored globally with a reference count and there is pointer to them in the process heaps.

Reply:
Monthly active users on github could be an interesting metric, because a lot of the code on github is web-related: https://www.benfrederickson.com/ranking-programming-languages-by-github-users/
TLDR: 1. JavaScript 2. Python 3. Java 4. PHP 5. C# 6.Go 7. TypeScript 8. Ruby 9. Jupyter Notebook 10. Objective-C 11. Swift 12. Kotlin 13. R 14. Scala 15. Lua 16. Matlab 17. CoffeeScript 18. Perl 19. Groovy 20. Haskell
I omitted languages that do not have a GC (C, C++, Rust) or that are clearly just there to implement command-line scripts (Shell, PowerShell).
CoffeeScript and Typescript can be collapsed into JavaScript.

Reply:
Speaking of "importance" of approaches of different languages, I'd like to point out a pretty strong trend to *transpile a given language down to C or LLVM IR or similar high-level representation*. Such language has more often than not a **garbage collector**. Such languages eagerly seek ways how to make them more performant in WASM runtimes without much effort. And **swapping** their garbage collectors for WASM GC (with some tiny shim layer) seems like a low hanging fruit.

I'm speaking of Nim, Go (cgo etc.), Rust, Crystal, etc. I think they're not that important, but it won't hurt to keep this approach in mind too while writing the requirements.

Reply:
These lists of programming languages from teh internets may give a *rough* impression of the landscape we are dealing with today, but how wise it is to prioritize on such a rough list that will be updated again next year and might look completely different in ten years? Shouldn't we instead prioritize on things we actually do know, like that JS will still be the primary interop target in browsers in ten years, that it is safe to assume that Java is the ideal candidate for Wasm GC (like C was the ideal candidate for the Wasm MVP), and that we also want to support functional languages well, just like there are tail calls for a reason, and that there'll likely be both nominal and structural type systems involved? An arbitrary listing will very likely just yield a never ending and perfectly justified stream of discussion (like this one) on the metrics used and, even worse, fuel disunion between the group, which we should strive to avoid.

Much of this thread then boils down to whether an individual language runtime may switch from RC or other 
more sophisticated (arena) collection strategies to a fire-and-forget kind of GC interface, which I'd assume the answer for is yes because it is much simpler, given that often necessary functionality like finalization can be provided. I'm certainly oversimplifying here, but how far this assumption does hold would be an interesting discussion to have.

Reply:
I think that there may be a misconception about having the explicit list of languages. We are not actually in the business of ranking these languages. Think of it as a criterion: if a language is not significantly popular, then it's demands are going to have to take a back seat relative to languages that are more popular. In most cases, this is unlikely to change the technical outcome; since at this level of abstraction, most languages have a lot in common (thanks Von Neumann).

Reply:
My argument was that this is unnecessary bike shedding involving risks for the community that go far beyond the questionably benefits of using a ranked list (something "We are not actually in the business of") as a criterion, and I am suggesting better metrics already available to us that can be used to avoid that. You ignored that part in its entirety, so I guess you agree?

Reply:
I do think that the main point is important: we need to focus on the popular languages.

Reply:
> I do think that the main point is important: we need to focus on the popular languages.

While not entirely false, it's rather dangerous to take that too far. It's at least as important that we don't overspecialise, don't prematurely optimise, and thereby lose the broader goal. Popularity is neither a scientific nor a stable criterion. Can we state a suitable requirement somehow without referring to dubious notions like that, or including an outdated list?

Reply:
I do believe it is extremely important to get some of the underrepresented language folks involved.

I don't believe WASM necessarily needs to prioritize features for popular languages, but it is a critical success factor to make sure they can run better on WASM than via translation to Javascript.

Currently, some of the most popular languages are underrepresented and, consequently, under-discussed. That obviously needs to change.

Reply:
Am including two sources of 'popularity'